src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -352,45 +377,54 @@ private AcceptTxResult AddCore(Transaction tx, TxFilteringState state, bool isPe
 
-                _transactions.TryGetBucketsWorstValue(tx.SenderAddress!, out Transaction? worstTx);
+                (tx.SupportsBlobs ? _blobTransactions : _transactions).TryGetBucketsWorstValue(tx.SenderAddress!, out Transaction? worstTx);
                 tx.GasBottleneck = (worstTx is null || effectiveGasPrice <= worstTx.GasBottleneck)
                     ? effectiveGasPrice
                     : worstTx.GasBottleneck;
 
-                bool inserted = _transactions.TryInsert(tx.Hash!, tx, out Transaction? removed);
-                if (inserted && tx.Hash != removed?.Hash)
-                {
-                    _transactions.UpdateGroup(tx.SenderAddress!, state.SenderAccount, UpdateBucketWithAddedTransaction);
-                    Metrics.PendingTransactionsAdded++;
-                    if (tx.Supports1559) { Metrics.Pending1559TransactionsAdded++; }
+                bool inserted = (tx.SupportsBlobs
+                    ? _blobTransactions.TryInsertBlobTx(tx, out Transaction? removed)
+                    : _transactions.TryInsert(tx.Hash!, tx, out removed));
 
-                    if (removed is not null)
-                    {
-                        EvictedPending?.Invoke(this, new TxEventArgs(removed));
-                        // transaction which was on last position in sorted TxPool and was deleted to give
-                        // a place for a newly added tx (with higher priority) is now removed from hashCache
-                        // to give it opportunity to come back to TxPool in the future, when fees drops
-                        _hashCache.DeleteFromLongTerm(removed.Hash!);
-                        Metrics.PendingTransactionsEvicted++;
-                    }
+                if (!inserted)
+                {
+                    Metrics.PendingTransactionsPassedFiltersButCannotReplace++;
+                    return AcceptTxResult.ReplacementNotAllowed;
                 }
-                else
+
+                if (tx.Hash == removed?.Hash)
                 {
-                    if (isPersistentBroadcast && inserted)
+                    if (isPersistentBroadcast)
                     {
                         // it means it was added and immediately evicted - we are adding only to persistent broadcast
                         _broadcaster.Broadcast(tx, isPersistentBroadcast);
                     }
                     Metrics.PendingTransactionsPassedFiltersButCannotCompeteOnFees++;
                     return AcceptTxResult.FeeTooLowToCompete;
                 }
+
+                (tx.SupportsBlobs ? _blobTransactions : _transactions).UpdateGroup(tx.SenderAddress!, state.SenderAccount, UpdateBucketWithAddedTransaction);
+                Metrics.PendingTransactionsAdded++;
+                if (tx.Supports1559) { Metrics.Pending1559TransactionsAdded++; }
+                if (tx.SupportsBlobs) { Metrics.PendingBlobTransactionsAdded++; }
+
+                if (removed is not null)
```

Comment:
should this go before `if (tx.Hash == removed?.Hash)` ?

Reply:
No, when `tx.Hash == removed?.Hash` it means that we added transaction and evicted it right after, so actually not added and need to return `FeeTooLowToCompete` at this point. So it's covering edge case scenario of adding transaction to the pool which is worse than any other tx in the pool

Later we are checking `removed is not null` and if it is true, it means that we added transaction and evicted other tx from the pool, so we need to invoke evicting event, remove this tx from cache to give it opportunity to come back and increment metric `PendingTransactionsEvicted`

That's perfectly fine in current shape

Reply:
Ok now it makes sense with how the code structure changed. Still could use better comments in code.